File: leetcode/107-二叉树的层序遍历 II/107-二叉树的层序遍历 II.go
package main

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 107-二叉树的层序遍历 II
func levelOrderBottom(root *TreeNode) [][]int {
	res := [][]int{}
	if root == nil {
		return res
	}

	queue := []*TreeNode{root}
	for len(queue) > 0 {
		tmp := []int{}
		size := len(queue)
		for i := 0; i < size; i++ {
			node := queue[0]
			queue = queue[1:]
			tmp = append(tmp, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		res = append(res, tmp)
	}

	l, r := 0, len(res)
	for l < r {
		res[l], res[r] = res[r], res[l]
		l++
		r--
	}
	return res
}

File: leetcode/107-二叉树的层序遍历 II/107-二叉树的层序遍历 II.js
/**
 * Definition for a binary tree node.
 * 
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

// 107-二叉树的层序遍历 II
var levelOrderBottom = function (root) {
    if (!root) { 
        return [];
    }
    const queue = [root];
    const res = [];
    while (queue.length > 0) { 
        let tmp = [];
        const size = queue.length;
        for (let i = 0; i < size; i++) { 
            node = queue.shift()
            tmp.push(node.val);
            if (node.left !== null) { 
                queue.push(node.left);
            }
            if (node.right !== null) { 
                queue.push(node.right)
            }
        }
        res.push(tmp);
    }
    res.reverse();
    return res;
};